roomserver/api/wrapper.go

```diff
@@ -47,34 +47,66 @@ func SendEventWithState(
 	ctx context.Context, rsAPI RoomserverInternalAPI, state *gomatrixserverlib.RespState,
 	event gomatrixserverlib.HeaderedEvent, haveEventIDs map[string]bool,
 ) error {
-	outliers, err := state.Events()
+	isCurrentState := map[string]struct{}{}
+	for _, se := range state.StateEvents {
+		isCurrentState[se.EventID()] = struct{}{}
+	}
+
+	authAndStateEvents, err := state.Events()
 	if err != nil {
 		return err
 	}
 
 	var ires []InputRoomEvent
-	for _, outlier := range outliers {
-		if haveEventIDs[outlier.EventID()] {
+	var stateIDs []string
+
+	for _, authOrStateEvent := range authAndStateEvents {
+		if authOrStateEvent.StateKey() == nil {
 			continue
 		}
-		ires = append(ires, InputRoomEvent{
-			Kind:         KindOutlier,
-			Event:        outlier.Headered(event.RoomVersion),
-			AuthEventIDs: outlier.AuthEventIDs(),
-		})
-	}
+		if haveEventIDs[authOrStateEvent.EventID()] {
+			continue
+		}
+
+		// Work out whether we should store the event as an outlier
+		// or if we should send it as a rewrite event.
+		storeAsOutlier := false
+		if authOrStateEvent.Type() == event.Type() && *authOrStateEvent.StateKey() == *event.StateKey() {
+			// The event is the same as the event that we are looking
+			// to send as a new event.
+			storeAsOutlier = true
```

Comment:
I don't grok why we therefore store this as an outlier.

Reply:
Well, there's three things that are given here:
- The event that we're sending (A)
- The state events (B)
- The auth events (C)

`authOrStateEvents` is B+C, so it's quite likely that A will replace something that's in B+C. If we don't store it as an outlier then we can't refer to it as an auth event.